src/nvim/buffer.c

```diff
@@ -3840,27 +3840,55 @@ int build_stl_str_hl(
     width = maxwidth;
 
   // If there is room left in our statusline, and room left in our buffer,
-  // add characters at the middle marker (if there is one) to
+  // add characters at the separate marker (if there is one) to
   // fill up the available space.
   } else if (width < maxwidth
-               && STRLEN(out) + maxwidth - width + 1 < outlen) {
-    for (int item_idx = 0; item_idx < itemcnt; item_idx++) {
-      if (item[item_idx].type == Middle) {
-        // Move the statusline to make room for the middle characters
-        char_u *middle_end = item[item_idx].start + (maxwidth - width);
-        STRMOVE(middle_end, item[item_idx].start);
-
-        // Fill the middle section with our fill character
-        for (char_u *s = item[item_idx].start; s < middle_end; s++)
-          *s = fillchar;
+             && STRLEN(out) + maxwidth - width + 1 < outlen) {
+    // Find how many separators there are, which we will use when
+    //  figuring out how many groups there are.
+    int num_separators = 0;
+    for (int i = 0; i < itemcnt; i++) {
+      if (item[i].type == Separate) {
+        num_separators++;
+      }
+    }
+
+    // If we have separated groups, then we deal with it now
+    if (num_separators) {
+      // Create an array of the start location for each
+      //  separator mark.
+      int separator_locations[STL_MAX_ITEM];
+      int index = 0;
+      for (int i = 0; i < itemcnt; i++) {
+        if (item[i].type == Separate) {
+          separator_locations[index] = i;
+          index++;
+        }
+      }
 
-        // Adjust the offset of any items after the middle
-        for (item_idx++; item_idx < itemcnt; item_idx++)
-          item[item_idx].start += maxwidth - width;
+      int standard_spaces = (maxwidth - width) / num_separators;
+      int final_spaces = (maxwidth - width) -
+        standard_spaces * (num_separators - 1);
+
+      char_u *sep_loc;
+      int dislocation;
+      int item_idx;
```

Comment:
Style: I'd move the above 3 declarations into the loop below and initialize them right away. Compilers will generate exactly the same code nowdays and it looks cleaner.